SplunkRumWorkspace/SplunkRum/SplunkRum/NetworkInstrumentation.swift

```diff
@@ -121,115 +121,109 @@ func startHttpSpan(request: URLRequest?) -> Span? {
     return span
 }
 
-class SessionTaskObserver: NSObject {
+class SessionTask: NSObject {
     var span: Span?
     // Observers aren't kept alive by observing...
-    var extraRefToSelf: SessionTaskObserver?
+    var extraRefToSelf: SessionTask?
     var lock: NSLock = NSLock()
     override init() {
         super.init()
         extraRefToSelf = self
     }
 
-    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey: Any]?, context: UnsafeMutableRawPointer?) {
-        lock.lock()
-        defer {
-            lock.unlock()
-        }
-        let task = object as? URLSessionTask
-        if task == nil {
-            return
-        }
-        if span == nil {
-            span = startHttpSpan(request: task!.originalRequest)
-        }
-        // FIXME possibly also allow .canceling to close the span?
-        if task!.state == .completed && extraRefToSelf != nil {
-            endHttpSpan(span: span,
-                        task: task!)
-            task!.removeObserver(self, forKeyPath: "state")
-            extraRefToSelf = nil
+    func networkSpan(task: URLSessionTask) {
+        DispatchQueue.main.async {
+            if self.span == nil {
+                self.span = startHttpSpan(request: task.originalRequest)
+            }
+
+            if  self.extraRefToSelf != nil {
```

Comment:
Are you sure it's not calling `endHttpSpan` right away after starting a span here? Have you checked the span durations? Since `extraRefToSelf` is assigned in init, this check is always true and the span is ended prematurely.

Besides, without KVO `extraRefToSelf` should no longer be necessary.

Reply:
Yes, I checked the span durations. then according to the response, endHttpSpan should be called?